internal/nexodus/policy_linux.go

```diff
@@ -0,0 +1,639 @@
+	}
+
+	// Create the egress nftables chains
+	if err := ax.nfCreateChain(egressChain); err != nil {
+		return fmt.Errorf("nftables setup error, failed to create nftables chain %s: %w", egressChain, err)
+	}
+
+	// Process the inbound rules
+	for _, rule := range inboundRules {
+		if len(rule.IpRanges) == 0 { // If the ip range is empty, add one
+			rule.IpRanges = append(rule.IpRanges, "")
+		}
+		if containsIPv4Range(rule.IpRanges) {
+			// if the rule is a L3 addresses in v4 family, with or without L4 port(s)
+			if err := ax.nfPermitProtoPortAddrV4(ingressChain, rule); err != nil {
+				return fmt.Errorf("nftables setup error, failed to process inbound v4 rule: %w", err)
+			}
+		} else if containsIPv6Range(rule.IpRanges) {
+			// if the rule is a L3 addresses in v6 family, with or without L4 port(s)
+			if err := ax.nfPermitProtoPortAddrV6(ingressChain, rule); err != nil {
+				return fmt.Errorf("nftables setup error, failed to process inbound v6 rule: %w", err)
+			}
+		} else if rule.FromPort != 0 && rule.ToPort != 0 {
+			// if the rule is L4 port(s) range with no l3 addresses
+			if err := ax.nfPermitProtoPort(ingressChain, rule); err != nil {
+				return fmt.Errorf("nftables setup error, failed to process inbound destination port rule: %w", err)
+			}
+		} else {
+			// if the rule is only protocol to permit (no L4 ports or L3 addresses)
+			if err := ax.nfPermitProtoAny(ingressChain, rule); err != nil {
+				return fmt.Errorf("nftables setup error, failed to process inbound destination port rule: %w", err)
+			}
+		}
+	}
+
+	// Process the outbound rules
+	for _, rule := range outboundRules {
+		if len(rule.IpRanges) == 0 { // If the ip range is empty, add one
+			rule.IpRanges = append(rule.IpRanges, "")
+		}
+		if containsIPv4Range(rule.IpRanges) {
+			// if the rule is a L3 addresses in v4 family, with or without L4 port(s)
+			if err := ax.nfPermitProtoPortAddrV4(egressChain, rule); err != nil {
+				return fmt.Errorf("nftables setup error, failed to process outbound v4 rule: %w", err)
+			}
+		} else if containsIPv6Range(rule.IpRanges) {
+			// if the rule is a L3 addresses in v6 family, with or without L4 port(s)
+			if err := ax.nfPermitProtoPortAddrV6(egressChain, rule); err != nil {
+				return fmt.Errorf("nftables setup error, failed to process outbound v6 rule: %w", err)
+			}
+		} else if rule.FromPort != 0 && rule.ToPort != 0 {
+			// if the rule is L4 port(s) range with no l3 addresses
+			if err := ax.nfPermitProtoPort(egressChain, rule); err != nil {
+				return fmt.Errorf("nftables setup error, failed to process inbound destination port rule: %w", err)
+			}
+		} else {
+			// if the rule is only protocol to permit (no L4 ports or L3 addresses)
+			if err := ax.nfPermitProtoAny(egressChain, rule); err != nil {
+				return fmt.Errorf("nftables setup error, failed to process inbound destination port rule: %w", err)
+			}
+		}
+	}
```

Comment:
There's a lot of duplication in here.

* The two loops appear to be the same other than which chain they're operating on.
* There's also a bunch of duplication between the `nfPermit*()` methods, especially `ProtoPortAddrV4()` and `ProtoPortAddrV6()`.

Now that you got all the rule types working and mostly covered with test cases, I think this could use some refactoring to simplify things by reducing duplication.

Reply:
Yeah, some of them are a bit duplicative between the address families (nfPermitProtoPortAddrV6 and nfPermitProtoPortAddrV4), part of the reason is they aren't very exchangeable. Particularly ICMP rules:

```
// nft add rule inet nexodus nexodus-outbound meta nfproto ipv6 ip6 daddr 2001:4860:4860::8888-2001:4860:4860::8889  iifname "wg0" accept
// // nft add rule inet nexodus nexodus-outbound meta nfproto ipv4 ip daddr 100.100.0.1-100.100.0.100 iifname wg0 accept
// nft add rule inet nexodus nexodus-inbound meta nfproto ipv6 ip6 nexthdr ipv6-icmp ip6 saddr 200::/64 counter accept
// nft add rule inet nexodus nexodus-inbound meta nfproto ipv4 ip protocol icmp ip saddr 100.100.0.0/20 counter accept
```

Where as some lend them self to be easily exchangable:

```
// nft add rule inet nexodus nexodus-inbound meta nfproto ipv4 tcp dport 0-65535 iifname "wg0" counter accept
// nft add rule inet nexodus nexodus-inbound meta nfproto ipv6 tcp dport 0-65535  iifname "wg0" counter accept
```

TLDR; the logic checks needed to happen somewhere, I chose to break it over more methods doing less, than overloading the methods with super long switch/case operations. Im still not positive all cases are captured, one thing I plan on doing next post-merge is to do a matrix of all possible combinations and make sure each is covered in e2e.

Reply:
OK, just leave it as is for now - thanks for the response